### ISSUE TITLE
OpenSSL >= 3.0.0 support

### ISSUE THREAD

**Issue by tlammi** [aws/aws-iot-device-sdk-cpp-v2#499]
### Describe the feature

This transition has to be done eventually anyway.

### Use Case

I'm developing an application which uses OpenSSL 3.0.,x but this started to crash due to segmentation faults a while back.
I tracked the issue and I think the bug is caused by libcurl and AWS SDK using different OpenSSL versions.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### SDK version used

v1.19.1, v1.19.3

### Environment details (OS name and version, etc.)

Manjaro Linux 22.0.0

**Comment by jmklix** [comment]
This is something that we are currently working on, but I don't have any detailed for when it will be released. The changes will be made in the aws-c-cal repo so I have opened a new issue there: https://github.com/awslabs/aws-c-cal/issues/132
Closing this issue

**Comment by graebm** [comment]
tldr; do a clean rebuilt of the IoT SDK, but this time pass `-DUSE_OPENSSL=ON` at the configure stage

Explanation:
- OpenSSL is composed of 2 main libraries: libcrypto (crypto math) and libssl (TLS state machine)
- The IoT SDK makes use of libcrypto, but not libssl
- The IoT SDK includes the source code for libcrypto, you can find it here: `crt/aws-crt-cpp/crt/aws-lc/crypto/...`
    - We do it this way to simplify the build process for anyone that doesn't already have openssl on their machine
- But your application is ALSO using libcurl, which is built to use the default libcrypto that's already installed on your system.
- Now you have an application that is trying to use 2 different libcryptos at the same time
- Some function calls go to 1 libcrypto, some function calls go to the other libcrypto, and that leads to explosions
- THEREFORE you need to build your application to only use ONE libcrypto
- You can force the IoT SDK to look for libcrypto on your system, instead of building its own copy, by configuring it with `-DUSE_OPENSSL=ON`

Does that fix your issue?

I got it working for me. I just tried building the IoT SDK against OpenSSL 3 (I used a docker image of Red Hat Enterprise Linux 9, because I googled for a distro that shipped with OpenSSL3 and that was the first hit). I configured with `-DUSE_OPENSSL=ON` and looking at the build output confirmed it was using the libcrypto from OpenSSL3. I built the MQTT5 pubsub sample and it works, no crashes.

**Comment by tlammi** [comment]
I already went another route and separated the AWS SDK to a separate binary, so I cannot really test this with the current setup. 

I compiled the SDK using the guide you gave me and everything compiled successfully. Unit tests and smoke tests also passed so at least the change did not break anything.

The update also fixed an issue where parts of the code were using OpenSSL 1 headers but linking against the system's OpenSSL 3 library. I guess this might have also contributed to the issue.
